Undefined variable T using generic on Class (PEP695)

### Bug description

```python
# pylint: disable=missing-module-docstring
class Simple[T]:
  def __init__(self, value: T):
    self.value = value
```


### Configuration

_No response_

### Command used

```shell
python3.12 -m pylint a.py
```


### Pylint output

```python
************* Module a
a.py:3:28: E0602: Undefined variable 'T' (undefined-variable)
```


### Expected behavior

There should be no error E0602 reported.

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.2
Python 3.12.1 (main, Dec  7 2023, 20:45:44) [Clang 15.0.0 (clang-1500.0.40.1)]
```


### OS / Environment

OSX - Sonoma 14.1.2 (23B92)

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/9335.false_positive
+++ b/doc/whatsnew/fragments/9335.false_positive
@@ -1,0 +1,4 @@
+Fix false positives for ``undefined-variable`` for classes using Python 3.12
+generic type syntax.
+
+Closes #9335

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1742,6 +1742,11 @@ class VariablesChecker(BaseChecker):
             if utils.is_ancestor_name(consumer.node, node) or (
                 not is_start_index and self._ignore_class_scope(node)
             ):
+                if any(
+                    node.name == param.name.name for param in consumer.node.type_params
+                ):
+                    return False
+
                 return True
 
             # Ignore inner class scope for keywords in class definition
@@ -1952,7 +1957,9 @@ class VariablesChecker(BaseChecker):
                 )
             return (VariableVisitConsumerAction.RETURN, found_nodes)
 
-        elif isinstance(defstmt, nodes.ClassDef):
+        elif (
+            isinstance(defstmt, nodes.ClassDef) and defnode not in defframe.type_params
+        ):
             return self._is_first_level_self_reference(node, defstmt, found_nodes)
 
         elif isinstance(defnode, nodes.NamedExpr):
@@ -2334,6 +2341,13 @@ class VariablesChecker(BaseChecker):
                     maybe_before_assign = defnode.value is node or any(
                         anc is defnode.value for anc in node.node_ancestors()
                     )
+                elif (
+                    isinstance(defframe, nodes.ClassDef)
+                    and defnode in defframe.type_params
+                ):
+                    # Generic on parent class:
+                    # class Child[_T](Parent[_T])
+                    maybe_before_assign = False
 
         return maybe_before_assign, annotation_return, use_outer_definition
 

--- a/tests/functional/g/generic_class_syntax_py312.py
+++ b/tests/functional/g/generic_class_syntax_py312.py
@@ -2,7 +2,7 @@
 class Entity[_T: float]:
     last_update: int | None = None
 
-    def __init__(self, data: _T) -> None:  # [undefined-variable]  # false-positive
+    def __init__(self, data: _T) -> None:
         self.data = data
 
 
@@ -28,6 +28,6 @@ class Parent[_T]:
         self.update_interval = 0
 
 
-class Child[_T](Parent[_T]):  # [undefined-variable]  # false-positive
+class Child[_T](Parent[_T]):
     def func(self):
         self.update_interval = None

--- a/tests/functional/g/generic_class_syntax_py312.txt
+++ b/tests/functional/g/generic_class_syntax_py312.txt
@@ -1,2 +1,0 @@
-undefined-variable:5:29:5:31:Entity.__init__:Undefined variable '_T':UNDEFINED
-undefined-variable:31:23:31:25:Child:Undefined variable '_T':UNDEFINED

--- a/tests/functional/u/undefined/undefined_variable_py312.py
+++ b/tests/functional/u/undefined/undefined_variable_py312.py
@@ -4,4 +4,5 @@ def f[T](a: T) -> T:
     print(a)
 
 class ChildClass[T, *Ts, **P]:
-    ...
+    def __init__(self, value: T):
+        self.value = value

